### DISCUSSION TITLE
Possible to pass state to next/router?

### DISCUSSION THREAD

**Discussion by bryan** [vercel/next.js#23991] [Ideas]
With react-router I was able to do: 

`browserHistory.push({ pathname: "/", state: { message: 'This is a message forwarded in a state.' } })`

Is it possible to do something like this using next/router or a way to redirect the user and also pass some value over to the new page?

**Comment by arunoda** [comment]
Currently that's not possible and I don't think we'll support it immediate future.

**Reply by oelbaga** [reply]
This really needs to be corrected. In react components represent pages and you can pass state between all components. Not being able to pass state between pages goes against the logic behind the React library.

**Reply by fultonm** [reply]
Next.js is not the same thing as React. Next.js is more focused on server side rendering from my understanding. It's great in some circumstances. The state you are mentioning - does it exist on the server as well as the client simultaneously? That's the problem at hand. Once I realized my app needed traditional React functionality (client-side focused) I went back to OG React.

**Reply by arturmuller** [reply]
I'd say Next.js need to explain the server/client issue in the docs as `History.pushState()` is a platform concept that you "loose" in Next. That said, I think it really isn't that much of an issue given that you can totally work around it using context, local or session storage, simple variables, etc.

**Reply by EloB** [reply]
@arturmuller The only thing that I agree on your statement is that Next.js disables the possibilities of using history state. It's not a easy task to store historic information in either context, local or session storage without it becoming a global state that often are really bad practice for UX.

Single/multiple element scrollbar positioning, collapsible, currentTime (video), form wizards states to name a few scenarios where historical states are way more suited than a global or component states.

Form wizards for instance are terrible with either global or component state.
- If you choose `global` state than you can't go back in history to an older filled version of the wizard because that will now contain the newer information from the latest filled wizard form. The only way to fix this is that you need to clean that global state after you submitted that form but then you will always come back to an empty form navigating with forward/back buttons in the browser. There are hacky ways to try to solve this by creating a "uuid" in the url when opening a wizard and keep the form data in the global state. The only problem doing so is that the global state will just grow and grow because it's hard to know when a historic event is no longer reachable.
- If you choose `component` state than you will lose all your information if that component gets unmount. Making the same behavior with reset on submit to the global state.

Both these practices are bad in general. The sad part is they are often used but I would say it's because of incompetence, to few blog posts on the topic and/or lack of support from frameworks.

**Reply by EloB** [reply]
I also want to say that history state is not a silver bullet. It also comes with it's challenges. For instance that you need "crypto" implementation to store sensitive wizard form data. So "hackers" can't extract previous form data from previous browser sessions if a user logged out.

This because browser lacks native api's to clean the history state. There are API's but they are only available for browser extensions: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/history/deleteAll.

**Reply by hirwaaldo1** [reply]
Having the ability to pass state in `useRouter()` would be a valuable addition.

**Comment by arunoda** [comment]
But let's open this and see whether we really really need this.

**Comment by nkzawa** [comment]
I think you should do this by passing query parameters.

**Reply by devodii** [reply]
really good fix for this.
but.. the problem becomes how to update the URL when it is no longer needed.

probably gonna incure a bunch of useEffects :D

**Comment by bryan** [comment]
I mean you can achieve a similar thing by passing in query parameters like nkzawa said, but I'd rather not pollute my the url with `http://www.example.com/?message=This%20is%20a%20long%20message%20forwarded%20to%20be%20displayed` and would rather have a state transfer like how react-router does it.

**Comment by rauchg** [comment]
You can use `as` to disguise the URL

**Comment by bryan** [comment]
Well I suppose so, but if I were to use `as`, the pathname doesn't retain the actual pathname in props.url.pathname, which means more things to pass into the query. It's just _cleaner_ to have it pass the state. 

I'm porting over a project to next.js and haven't run into any major compatibility problems since I found ways to refactor codes to make it work, it's just inconvenience and verbosity at this point that can be improved and all this issue is.

Thanks all for your input.

**Comment by timneutkens** [comment]
Closing this cause of inactivity 👍

**Comment by bryan** [comment]
A little over a year since I brought this up, but is this still not supportable?

**Comment by sergiodxa** [comment]
You can save any required state in localStorage, sessionStorage, a variable, a Redux store, etc. and then use it in getInitialProps (checking you are running client side).

**Comment by egaleme** [comment]
i think nextjs should start supporting react-router

**Comment by vaibhavi3t** [comment]
I want same functionality.

Earlier I was using CRA and there I used to send state to routed component. I wants same functionality with nextjs. 

What @sergiodxa suggested in workaround but library support will be good.

**Comment by todesstoss** [comment]
Same here, would be great if it will be possible pass data with `Router.push()` and retrieve it later with `withRouter`

**Comment by namy16** [comment]
I want same functionality. Any update on this ?

**Comment by namlq93** [comment]
> I want same functionality. Any update on this ?

me too

**Comment by allicanseenow** [comment]
Another question related to this issue is:
- I don't want to pollute my URL when I call the Router.push() and I want to pass a state to the next component. Is there any way to do it? Using localStorage or cookie is a bad practice as they are not directly controlled by React hooks or the old React lifecycle

**Comment by abdulhadi1995** [comment]
I have a scenario where I have a clickable component. clicking it would open its details page. With current next.js i can pass the projects id and then get the details using useEffect and axios call but since i already have the data i really wished i could send the state to that page.

**Comment by avin-kavish** [comment]
Would be nice to have this.

@sergiodxa Yes, there are many ways to pass temporary messages but the advantage of having it tied to the router is that we don't have to write extra code to manage the life cycle of the values we pass around. They will be tied to router navigation events and unavailable in other contexts.

For example, If I push it to my redux store, then I have to exclude it from being persisted to local storage and also dispatch a clear message action on the receiving end.

If I add it the query string it becomes a perma link that uses can bookmark and later return to.

But it I can push it to the next page and only the next page via `router.push` I don't have to make any compensations; the value will only be available in that context.

Very similar to [connect-flash](https://github.com/jaredhanson/connect-flashl)

**Comment by theoludwig** [comment]
No solution since Jan 15 2017 ?
We need that feature : 
> Same here, would be great if it will be possible pass data with `Router.push()` and retrieve it later with `withRouter`

**Comment by davidebukali** [comment]
Would be nice to have this in nextjs

**Comment by azamiya** [comment]
if you want to hide query params, just use like this

```
Router.push(`/main?userName=${userName}`, 'main')
```

than the main page, you can see only `/main` on url and
`console.log(Router.route)` shows query

```
pathname: "/main"
query: {userName: "fadsfasd"}
asPath: "main"
```

**Comment by avin-kavish** [comment]
oh yeah, that's true. Btw I find it really awkward that router requires two strings. `react-router` works beautifully without, why does this?

**Comment by batistafrancis** [comment]
For those who want to make it, you can use `Context Api`.

Next.js provides the `router.push()` passing `shallow: true` as a prop.
It's the most simple state management to do that.
https://nextjs.org/docs/api-reference/next/router#routerpush

**Comment by timneutkens** [comment]
Willing to reconsider this, however it needs a **great** proposal on how to prevent users from shooting themselves in the foot using history state given you can't access history state server-side for example, it would be very easy to cause hydration issues.

Do note that re-opening this does not mean that I'm not going to write a RFC for this feature, that would be up to the people that want to have this feature.

**Comment by myoneflag** [comment]
`
    Router.push({
      pathname: '/',
      query: { message: 'This is a message forwarded in a state.' }
    }, '/')`

**Comment by umakantp** [comment]
More than 2 years and still waiting for state feature

**Comment by avin-kavish** [comment]
> ` Router.push({ pathname: '/', query: { message: 'This is a message forwarded in a state.' } }, '/')`

This is still not internal state, it's a query string with a mask feature, which is quite an odd thing to do in my opinion.

**Comment by pxwee5** [comment]
I'm not super familiar with React ecosystem. But it appears that you can do
```
router.push(
  { pathname: '/hello-world' }, 
  '/hello-world/2', 
  { step: 2 }
);
```

When you go back and check before PopState, the option prop will contain `{ step: 2 }`. 

We only need this in the `router` object.

**Comment by likerRr** [comment]
> I'm not super familiar with React ecosystem. But it appears that you can do
> 
> ```
> router.push(
>   { pathname: '/hello-world' }, 
>   '/hello-world/2', 
>   { step: 2 }
> );
> ```
> 
> When you go back and check before PopState, the option prop will contain `{ step: 2 }`.
> 
> We only need this in the `router` object.

It seams like another temporal hack, not intentional behavior. Expected options are listed in the [docs](https://nextjs.org/docs/api-reference/next/router#routerpush). But something similar to this would be great to have as expected behavior

**Comment by pxwee5** [comment]
@likerRr Apparently all of the options parameter is being passed to `window.history.state`. So with my example:
`window.history.state == { step: 2 }`

**Comment by likerRr** [comment]
> @likerRr Apparently all of the options parameter is being passed to `window.history.state`. So with my example:
> `window.history.state == { step: 2 }`

Seems like it is. Anyway, until it's documented, it can't be used in prod, because implementation of passing options can be changed anytime :disappointed: Another problem (?) with history state, that it's persistent. And if you play with "back"/"forth" buttons you will mostly get unexpected behavior.

As far as I understand the use case of what community wants is to pass data which will leave only during transition from one page to another and we don't have to bother of clearing it. And this data should not appear again when user clicks "back/forth" in a browser. Otherwise [this](https://github.com/vercel/next.js/issues/771#issuecomment-361038497) is the best proposition (imo).

**Comment by ken-nah** [comment]
Would be nice if next js considered this

**Comment by TrongTrinh** [comment]
This worked for me, you can use
https://github.com/vercel/next.js/issues/771#issuecomment-612018764

1: in the Subscription page: 
const handleClick = useCallback((mount) => {
        Router.push(`/payment?mount=${mount}`, '/payment');
    }, []);
2: in Payment Page
useEffect(() => {
        const {
            query: { mount },
        } = Router;
        !mount && Router.push('/campaign-overview/subscription');
        console.log('PaymentVM:React.FC -> mount', mount);
    }, []);

you can see value of `mount` in console browser.

**Comment by meilechwieder** [comment]
I understand the issue and the convenience of being able to  pass along state in router.push. 

Would the following be a temporary solution?

Use the browser's history.pushState, as well as a local useState of a location string, and when we want to redirect, we use history.pushState as well as setting a new value to local location state. An update to the location-string-state should trigger a useEffect run. Then have a JS swich statment which evaluates the location and returns components based on the location. And if we wanna get to the same component without history.pushState (for exp. if the user refreshes the page), we can make the location a dynamic next query param. -- did not try it, just trying to think of workarounds.

**Comment by meilechwieder** [comment]
> For those who want to make it, you can use `Context Api`.
> 
> Next.js provides the `router.push()` passing `shallow: true` as a prop.
> It's the most simple state management to do that.
> https://nextjs.org/docs/api-reference/next/router#routerpush

What if in my in my `getServerSideProps`, I make two calls, one of which should be redone for every load, and one should only be loaded with a condition. For that, `Shallow: true` would not help much because its not a state manager, its either redo `getServerSideProps`, or don't do it at all.
An example of this scenario would be if I make two seperate calls in `getServerSideProps`. One for the main content of the page and then another call to get the navigation links. In an initial render, I want both calls to be on the server, but then, I already have the nav links in memory so I dont need to refetch them. Thats why we need state. 

The only solution is to send along a `as` option with hidden query params but our code would be awful it for every `<Link/>` we pass in a bounch of query params that contains our current state. (Actually, as Im writing, Im thinking of making an HOC Link which should go the the specified href and send along a query object with the current state so I wouldnt need to write the state object in each `<Linke/>`, I'll only handle this in the HOC)

**Comment by a-marcel** [comment]
Hello, 

are there any explanations from the vercel / next team, why it would be problematic to pass a state to the router/link ? It looks like this topic is required by a lot of people (even from me - and i don't see any valid reason against this in this chat).

My usecase is tostore the scroll position (not of the page, the scroll position from an overflow:hidden div), so that a user can go back and have the same position inside this div as before. 

I could imaging to have a[ state property](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/client/link.tsx#L21) for the [NextLink](https://nextjs.org/docs/api-reference/next/link).

`<Link ... sate={{ value: 1 }}>LInk that store the State </Link>
`

This param can be passed to the [router push](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/client/link.tsx#L145). The next framework is using already the state for their purpose.

I and mostly all in this thread would appreciate if request can be added to the code. 

Thanks a lot for all your nice work and the nextjs framework

**Reply by oelbaga** [reply]
This is SOOOO needed. Im having a difficult time seeing the Next Js framework usable for web without this feature. I didn't realize you couldn't pass state or any data in between pages (outside of url querystrings) until I started building an app with Next JS and my jaw dropped.

There seems to be a disconnect with NextJs on how we require data to be passed in between pages on the web just like it's passed between components in React.

**Comment by rodbs** [comment]
I'd also be interested in having this functionality of passing state to next/router. Thx!

**Comment by marcin-kochanek** [comment]
Me too!

**Comment by timneutkens** [comment]
As said earlier: https://github.com/vercel/next.js/issues/771#issuecomment-622362614

**Comment by a-marcel** [comment]
Okay, I agree, this is a valid concern.

**Comment by meilechwieder** [comment]
> Hello,
> 
> are there any explanations from the vercel / next team, why it would be problematic to pass a state to the router/link ? It looks like this topic is required by a lot of people (even from me - and i don't see any valid reason against this in this chat).
> 
> My usecase is tostore the scroll position (not of the page, the scroll position from an overflow:hidden div), so that a user can go back and have the same position inside this div as before.
> 
> I could imaging to have a[ state property](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/client/link.tsx#L21) for the [NextLink](https://nextjs.org/docs/api-reference/next/link).
> 
> `<Link ... sate={{ value: 1 }}>LInk that store the State </Link> `
> 
> This param can be passed to the [router push](https://github.com/vercel/next.js/blob/bbc1a21c749c423e842586ab116889c9f9c7024e/packages/next/client/link.tsx#L145). The next framework is using already the state for their purpose.
> 
> I and mostly all in this thread would appreciate if request can be added to the code.
> 
> Thanks a lot for all your nice work and the nextjs framework

Actually, you can do it with hidden query params. In nextjs you can redact query params with the `as` prop on the `<link/>` component or with the second argument for the `router.push` method.

**Comment by a-marcel** [comment]
@meilechwieder : Thanks for this idea. It's not a clean solution but with a dedicated React  Link Component (only for this scrolling case) it could work out without creating so much dept.

**Comment by sughodke** [comment]
Workaround: You can use HTML5 `replaceState` to clean up the window location.

```
history.replaceState(null, '', `/post/${id}`);
```

https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState

**Comment by mv29** [comment]
I am still waiting for this functionality, hope it gets implemented in near future.

**Comment by fultonm** [comment]
Yeah this feature is really needed. The hidden query params work around is nice in the mean time, but we may need to pass complex objects to the next page.

**Comment by ATakaSKY** [comment]
How can I do the same thing via Link tag. Currently how i pass data is like this:
```
<Link
    href={{
      pathname: url,
      query: { a: 123 }
    }}
    onClick={cardClicked}
>Card</Link>
```

But this results in appending query params on navigation, is there a way to hide query params like we do in router push by passing '/' as the second parameter.

**Comment by meilechwieder** [comment]
Sure, the as prop:

<Link
    href={{
      pathname: url,
      query: { a: 123 }
    }}

    as={{

       pathname: '/',

       query: {}

    }}
    onClick={cardClicked}
>Card</Link>

**Comment by jochasinga** [comment]
Instead of polluting the router, which appears to be stateless by design (as in REST), is it possible to leverage React's `Context` for a global state that's decoupled from routing?

For instance, for a fictional `User` state, I could create a higher-order component like so:

```javascript
// store.js

import { createContext, useReducer } from 'react';

let initState = {
  name: '',
  age: 0,
};

const UserStateContext = createContext({state: initState});

const UserProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initState)
  return (
    <UserStateContext.Provider value={{state, dispatch}}>
      {children}
    </UserStateContext.Provider>
  )
}

export { UserStateContext, UserProvider };
```

At the level I'd like to keep that state global, let's say app level, I then wrap it like so:

```javascript
import { UserProvider } from '../store'

function App({ Component, pageProps }) {
  return (
    <UserStateProvider>
      <Component {...pageProps} />
    </UserStateProvider>
  )
}

export default App;
```

This way, from any child component one could easily access the current user state (and its dispatch function) with `useContext(UserStateContext)`. This could be a nice workaround and it's already there in React. I find it cleaner than passing state into the router context.

**Comment by kimbaudi** [comment]
couldn't we prevent polluting the url by using `router.replace`?

For my use case, I use `as` prop to mask the query and also prevent polluting the url using `router.replace`:

my LogoutButton component that is in all pages:

```jsx
const handleLogout = () => {
  router.push({ pathname: "/", query: { action: "SIGN_OUT" } }, "/");
};
```

my PageWrapper component that wraps all pages:

```jsx
useEffect(() => {
  const { query } = router;

  if (query.action === "SIGN_OUT") {
    authProvider()
      .signOut()
      .then(() => {
        router.replace("/");
      })
      .catch((error) => {
        // handle error
      });
  }
}, [authProvider, router]);
```

**Reply by deiga** [reply]
If you're sending actions around would it make sense to use a reducer and dispatch instead?

**Comment by oelbaga** [comment]
I was searching for the same thing. Honestly this makes no sense at all to not be supported, because in React you can do it by navigating from component to component, but in Next we use pages for navigation which is great yet you're stripping away the core functionality of being able to pass data to pages outside of a query params.

You can't pass multiple pieces of state data in a query parameter because you need a url to be SEO clean that has nothing to do with that data. State data from one page to another page should not be passed through the URL Query String anyway.

The only other option is to somehow re-call all of our state and API data on each page call through a single keyword in the URL. 

This is is a critical feature in truth. I'm pretty disappointed after hitting this roadblock while testing our NextJs for an upcoming web application.

**Reply by Ananthakrishnanj** [reply]
Absolutely agree to your point

**Comment by detoner777** [comment]
Still really needs this!

**Comment by MeLlamoPablo** [comment]
Is Vercel interested in accepting pull requests for this feature?

**Comment by benjipott** [comment]
you can use :

```
const { push } = useRouter()

push(path, undefined, { shallow: true, toto: true })
```

and in your component :

```
export const useLocationState = (): LocationState => {
  const { asPath, query } = useRouter()

  const state = typeof window !== 'undefined' && window.history?.state?.options

  return state
}
```

**Comment by EloB** [comment]
Using only component state is terrible for UX. If a user want to go back some pages and suddenly all component state are reset or having a global state that changes all pages is not good. Very common these days that you fill a form and click back button to see that all your form data is gone because they were component states.

For instance you have some accordions. You don't want those state (open or close) in your global state. That is something that you only want to be on that specific time in history.

Video currentTime (to be able to continue where you were after a page refresh, F5), scroll position on elements, form wizards are some good use cases too.

I often use my history state as a reducer (immutable) and on pushState I start from a clean state, might be different depending on context.

**Comment by danielweil** [comment]
Would love the option to use this feature, instread of limiting this capability

**Comment by radikrisffnext** [comment]
I can also emphasize, that this feature would be really useful and make many webapps more efficient!

**Comment by ghost** [comment]
I love next js but 5 years later this hasn’t even been looked. Disappointing.

**Comment by magoz** [comment]
This feature would be super useful!

**Comment by wparam** [comment]
It's nice to have this feature, looking up to it.

**Comment by omerman** [comment]
Could we get any explanations as to why 2022 and we don't have  an equivalent method for history.pushState that actually accepts a state? 😶

**Comment by OscarVantagePOS** [comment]
I use this for things such page state (such as whether a modal is open) as it doesn't really matter if the state gets lost and the URL stays tidy at all times. Also useful for storing breadcrumbs as past history can't be accessed so I just pass a log of where the user has been so they can navigate easily, even after a refresh.

I know this has been mentioned a million times but stuff it I'm saying it too. This is supported by the browser history API already (https://developer.mozilla.org/en-US/docs/Web/API/History/pushState).

To do this currently requires adding the state to the options object and it seems that it will pass that into the state.
For example:
```js
router.push(
	'/product',
	undefined,
	{
		state: {
			modalOpen: false
		}
	}
);
```
It can be accessed from history directly through `history.state.options.state.modalOpen`.

And even this doesn't work as it is overridden when the page refreshes.

**Comment by muhayli** [comment]
NEXT JS should really add this feature it's really important
but one sloution that works for me is to pass query object and parse any object to json and when retrieving it stringify that object i.e

WHEN PUSHING STATE:
    router.push({
      pathname: "/hotels",
      query: {
        destination,
        options: JSON.stringify(options),
        date: JSON.stringify(date),
      },
    });



WHEN RETREVING STATE INTO USESTATE:
  const [destination, setDestination] = useState(
    (router.query.destination as string) ?? ""
  );
  const [date, setDate] = useState<Range[]>(
    router.query.date
      ? JSON.parse(router.query.date as string)
      : [{ startDate: new Date(), endDate: new Date() }]
  );
  const [options, setOptions] = useState<OptionsType>(
    router.query.options
      ? JSON.parse(router.query.options as string)
      : {
          adults: 0,
          children: 0,
          room: 0,
        }
  );

**Comment by LeulAria** [comment]
So what happend to this issue 💀

**Reply by dejan966** [reply]
I would like to know too. It's 2024 and it seems like there's no permanent workarund

**Comment by OscarVantagePOS** [comment]
I found a funny way of doing it. Not a great workaround but it works for my purposes:
```
import { useEffect, useState } from 'react';

export default function useHistoryState(state: any, length: number, updateState: (state: unknown) => void) {
	const [hydrated, setHydrated] = useState(false);

	useEffect(() => {
		const parsedSessionState = JSON.parse(sessionStorage.getItem('historyState') ?? '{}');

		if (parsedSessionState) {
			const largestKey = Math.max(...Object.keys(parsedSessionState).map(el => +el));

			if (length in parsedSessionState) {
				updateState(parsedSessionState[length]);
			} else if (length > largestKey) {
				updateState(parsedSessionState[largestKey]);
			}
		}

		setHydrated(true);
	}, []);

	useEffect(() => {
		if (hydrated) {
			const sessionState = JSON.parse(sessionStorage.getItem('historyState') ?? '{}');

			sessionState[length] = state;

			sessionStorage.setItem('historyState', JSON.stringify(sessionState));
		}
	}, [state, length, hydrated]);
}
```
I use react-router, though I'll be changing to the NextJS router in the future. All you need to do is pass in the current history state, history.length, and a function to reroute when the hook tries to hydrate the state. I use history.replace with the current location pathname.
With the nextjs router you can technically pass in state but you'll need to use ts-ignore.

Hopefully this helps someone. I am hoping to make a PR in the future that introduces history.state functionality in a way that works with NextJS rather than hacking it.

**Reply by maranomynet** [reply]
In fact Next.js's router sets `window.history.state` to be an object that has a unique (hash) `key` property that can be used as a key for `sessionStorage`.
This can be abused to do all sorts of fun things!

**Comment by jlevy-io** [comment]
Migrating a large Gatsby project to Next.js and this is the biggest bump in the road I've encountered so far.   We utilize react-router's history state for allowing users to use native browser controls (back, forward, etc) to navigate through form steps on a single route.   I would love to see similar functionality with Next router

**Comment by urielaweobe** [comment]
My solution

```
import { useMemo } from "react";
import { useRouter } from "next/router";
import { data } from "../../utils";

const Example = () => {
    const router = useRouter();
    const { pid } = router.query;

    const details = useMemo(() => {
        return data.find((item) => item.id === Number(pid));
      }, [pid]);

    return (
          <div>
            <p>{details?.name}</p>
          </div>
      );
};

export default Example;
```
Worked for me. 
"next": "13.1.1";

**Reply by dejan966** [reply]
From where is pid coming from and how did you pass it here?

**Comment by ramirez** [comment]
So, is this issue even being considered?

**Reply by muhayli** [reply]
Since that is how nextjs functions and this is not considered a bug or anything similar, I don't believe it will be considered.

**Comment by EloB** [comment]
An bad idea but probably the only way to get around this annoying issue. Is to override the `history` object and these methods. Than inject this before all client-side code from Next.js being initiated.

This is just pseudo code needs also injecting into `popstate`.
```ts
const methods = ["back", "forward", "go", "pushState", "replaceState"];
const proxy = Object.fromEntries(methods.map((key) => [key, history[key]]))

history.back = function back(...args) {
  // modify args...
  return proxy.back.apply(this, args);
};
history.forward = function forward(...args) {
  // modify args...
  return proxy.forward.apply(this, args);
};
history.go = function go(...args) {
  // modify args...
  return proxy.go.apply(this, args);
};
history.pushState = function pushState(...args) {
  // modify args...
  return proxy.pushState.apply(this, args);
};
history.replaceState = function replaceState(...args) {
  // modify args...
  return proxy.replaceState.apply(this, args);
};
```

Keep Next.js history states seperate. Then we should be able to at least keep state in history.

```ts
history.state: {
  nextState: {
    // here is namespace for all next.js history states.
  },
  appState: {
    // here is namespace for your app history states.
  }
}
```

If there are some helper functions that prevents app history state only to affect that app state than it should be fine.

References:
https://developer.mozilla.org/en-US/docs/Web/API/History_API

**Comment by jscul** [comment]
It would be fine if I could use Link `as` prop... but that doesn't seem to be working in Next 13.4.18.

```
<Link
  href={{
    pathname: `/content/${i}/post/${i}`,
    query: { relatedPost: true },
  }}
  key={i}
  as={`/content/${i}/post/${i}`}
>
```

^ I'm not able to get the query param relatedPost on `/content/${i}/post/${i}`.

**Comment by sandeepbfnp** [comment]
what if i am redirecting my page to new page on new tab still it will work or not

**Comment by JayJ-WeCode** [comment]
any solution? 💀

**Comment by daxdesai** [comment]
This is working
`history.push({
                                        pathname: '/results',
                                        state: {
                                            pattern: value,
                                        }
                                    });`

**Reply by QYunChao** [reply]
I try `history.push` but get `Uncaught TypeError: history.push is not a function`, can you tell me where the `history` from, thanks

**Reply by ekebobrown** [reply]
I want to believe that happened on a client component as **history** is a browser API
[https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#windowhistorypushstate](url)

**Comment by benwiley4000** [comment]
This basically makes Next router a toy

**Comment by BenEinwechter** [comment]
With the AppRouter, you can now use native `history.pushState` and `history.replaceState` https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api 

Looks like they added experimental support for using the native history API in v14.0.3 and I'm guessing it landed as stable sometime after that since the docs don't say it's experimental. Here's the original PR https://github.com/vercel/next.js/pull/58335 so you can read through it to see if any of the tradeoffs affect you.

**Comment by benwiley4000** [comment]
Thanks @BenEinwechter. I tested this feature out but it seems pretty useless since for client-side navigation it always overwrites history.state immediately. It ought to extend the user-supplied history.state rather than replacing it.

**Reply by Dinqe3** [reply]
Did u find any solution to this? Same problem here @benwiley4000

**Comment by benwiley4000** [comment]
No, I wasn't able to find any workaround except for forcing my state into
the URL, which is recommended by next I guess, but I don't consider this a
good solution.

Le mer. 13 mars 2024, 5 h 21 a.m., danbru ***@***.***> a
écrit :

**Reply by Dinqe3** [reply]
Okay. Me either. Found this but did not give it a try: https://github.com/vercel/next.js/pull/58335

Solved it by using SessionStorage in the meanwhile.

**Comment by lockehooper** [comment]
I ran into this issue today and wanted to share a more "Reacty" work around I ended up going with that allows for passing around arbitrarily complex objects. The basic idea is to use the React context to pass whatever objects we want around. The full code for what I did is [here](https://github.com/lockehooper/next-router-parms).

For those facing this problem looking for a solution that uses traditional React tools/hooks, the context provider is below:
```
export const LazyDataStoreProvider = ({ children }: { children?: React.ReactNode }) => {
	const internalMap = new Map<string, Object>();

	const contextValue = useMemo(() => {
		return {
			getStoredValue: <T extends Object = Object>(key: string, fetch?: () => T, fallback?: () => void): T | undefined => {
				const foundValue = internalMap.get(key);
				if (foundValue) {
					return foundValue as T;
				}
				if (fetch) {
					const fetchedValue = fetch();
					internalMap.set(key, fetchedValue);
					return fetchedValue;
				}
				if (fallback) {
					fallback();
				}
				return undefined;
			},
			setStoredValue: (key: string, value: Object) => {
				internalMap.set(key, value);
			},
			deleteStoredValue: (key: string) => {
				internalMap.delete(key);
			},
		};
	}, []);

	return <LazyDataStore.Provider value={contextValue}>{children}</LazyDataStore.Provider>;
};
```

I added this provider at ```app/layout.tsx``` (although it doesn't really matter as long as it wraps both the to and from page). Then call the ```setStoredValue``` method right before ```router.push```. Then I can use the id that is passed as part of the URL to fetch it from the next page using ```getStoredValue```. I also included ```fetch``` and ```fallback``` methods that can be used to avoid the user getting stuck if the value they need isn't stored (such as if the page is reloaded). This (or something similar like a bookmark at "/page/[id]") is a common problem case I built a hook for where the backup behavior is to redirect the user to a provided URL:

```
export const useStoredOrRoute = <T extends Object = Object>(key: string, url: string): T | undefined => {
	const router = useRouter();
	const { getStoredValue } = useLazyDataStore();
	const [shouldRoute, setShouldRoute] = useState(false);

	useEffect(() => {
		if (shouldRoute) {
			router.push(url);
		}
	}, [shouldRoute]);

	return useMemo(() => getStoredValue<T>(key, undefined, () => setShouldRoute(true)), [key]);
};
```

**Comment by PunkFleet** [comment]
I am having the same problem and am migrating from react-router. useRouter as a client component should have state functionality and be client focused.

**Comment by benwiley4000** [comment]
If you have an option to stick with react router I wouldn't migrate to next
router, especially if you're relying on history state.

Le lun. 18 mars 2024, 4 h 10 a.m., birddle ***@***.***> a
écrit :

**Comment by Nura-Programmer** [comment]
Does this means we will have to impose redux for passing a single state?

**Reply by benwiley4000** [reply]
This can work in some cases but will not replace the full functionality of history-bound state. Particularly since history state persists after a refresh.

**Reply by Nura-Programmer** [reply]
I just render it on the page replacing the form as soon as the response arrived

**Comment by ghost** [comment]
Bhai upi krta 100 INR, krde yaar

**Comment by Shivender-Kun** [comment]
Any updates???

**Comment by olamilly** [comment]
Really need this

**Comment by bloodykheeng** [comment]
any updates on this

**Comment by ninjaguydan** [comment]
Also begging for this in 2026. Guessing it's a no-go so far

**Reply by bloodykheeng** [reply]
I think it wont be possible just use queryParams
